packetery/src/Packetery/class-checkout.php

```diff
@@ -0,0 +1,236 @@
+		foreach ( $chosen_methods as $method ) {
+			if ( strpos( $method, 'packetery_shipping_method' ) !== false ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Renders widget button and information about chosen pickup point
+	 */
+	public function render_widget_button(): void {
+		if ( self::is_packetery_order() ) {
+			$language = substr( get_locale(), 0, 2 );
+
+			$country = strtolower( WC()->customer->get_shipping_country() );
+			if ( ! $country ) {
+				$country = strtolower( WC()->customer->get_billing_country() );
+			}
+			if ( ! $country ) {
+				$this->latte_engine->render( PACKETERY_PLUGIN_DIR . '/template/checkout/error_country.latte' );
+
+				return;
+			}
+
+			$weight = WC()->cart->cart_contents_weight;
+			$weight = wc_get_weight( $weight, 'kg' );
+
+			$this->latte_engine->render(
+				PACKETERY_PLUGIN_DIR . '/template/checkout/widget_button.latte',
+				array(
+					'language' => $language,
+					'country'  => $country,
+					'weight'   => $weight,
+				)
+			);
+		}
+	}
+
+	/**
+	 * Renders main checkout script
+	 */
+	public function render_after_checkout_form(): void {
+		$app_identity = 'woocommerce-' . get_bloginfo( 'version' ) . '-' . WC_VERSION . '-' . Plugin::VERSION;
+
+		$this->latte_engine->render(
+			PACKETERY_PLUGIN_DIR . '/template/checkout/checkout_script.latte',
+			array(
+				'app_identity'       => $app_identity,
+				'pickup_point_attrs' => self::$pickup_point_attrs,
+				'packetery_api_key'  => get_option( 'packetery_api_key' ),
+			)
+		);
+	}
+
+	/**
+	 * Adds fields to checkout page to save the values later
+	 *
+	 * @link https://docs.woocommerce.com/document/tutorial-customising-checkout-fields-using-actions-and-filters/
+	 *
+	 * @param array $fields Fields before adding.
+	 *
+	 * @return array
+	 */
+	public static function add_pickup_point_fields( array $fields ): array {
+		foreach ( self::$pickup_point_attrs as $attr ) {
+			$fields['shipping'][ $attr['name'] ] = array(
+				// You can use label key for other types.
+				'type'     => 'hidden',
+				'required' => false,
+			);
+		}
+
+		return $fields;
+	}
+
+	/**
+	 * Checks if all pickup point attributes are set, sets an error otherwise.
+	 */
+	public static function validate_pickup_point_data(): void {
+		if ( self::is_packetery_order() ) {
+			if (
+				! isset( $_POST['_wpnonce'] ) ||
+				! wp_verify_nonce( sanitize_text_field( wp_unslash( $_POST['_wpnonce'] ) ), self::NONCE_ACTION )
+			) {
+				wp_nonce_ays( '' );
+			}
+
+			// TODO: validate carrier_id and point_carrier_id.
+			$required_attrs = array_filter(
+				array_combine(
+					array_column( self::$pickup_point_attrs, 'name' ),
+					array_column( self::$pickup_point_attrs, 'required' )
+				)
+			);
+			foreach ( $required_attrs as $attr => $required ) {
+				$attr_value = null;
+				if ( isset( $_POST[ $attr ] ) ) {
+					$attr_value = sanitize_text_field( wp_unslash( $_POST[ $attr ] ) );
```

Comment:
https://rules.sonarsource.com/php/tag/bad-practice/RSPEC-2043
Při unit testech je potřebuješ měnit, ale mají být read only. Doporučuji vytvořit službu \Nette\Http\Request.